desktop_version/src/Music.cpp

```diff
@@ -13,160 +14,513 @@
+#define sin(x) SDL_sin(x)
+#define cos(x) SDL_cos(x)
+#define floor SDL_floor
+#define abs(x) SDL_abs(x)
+#define ldexp(v, e) SDL_scalbn((v), (e))
+#define exp(x) SDL_exp(x)
+
+#define qsort SDL_qsort
+
+#define assert SDL_assert
+
+#define FILE SDL_RWops
+#ifdef SEEK_SET
+#undef SEEK_SET
+#endif
+#ifdef SEEK_CUR
+#undef SEEK_CUR
+#endif
+#ifdef SEEK_END
+#undef SEEK_END
+#endif
+#ifdef EOF
+#undef EOF
+#endif
+#define SEEK_SET 0
+#define SEEK_CUR 1
+#define SEEK_END 2
+#define EOF -1
+#define fopen(path, mode) SDL_RWFromFile(path, mode)
+#define fopen_s(io, path, mode) (!(*io = fopen(path, mode)))
+#define fclose(io) SDL_RWclose(io)
+#define fread(dst, size, count, io) SDL_RWread(io, dst, size, count)
+#define fseek(io, offset, whence) SDL_RWseek(io, offset, whence)
+#define ftell(io) SDL_RWtell(io)
+
+#define FAudio_alloca(x) SDL_stack_alloc(uint8_t, x)
+#define FAudio_dealloca(x) SDL_stack_free(x)
+
+#define STB_VORBIS_NO_PUSHDATA_API 1
+#define STB_VORBIS_NO_INTEGER_CONVERSION 1
+#include <stb_vorbis.h>
+
+/* End stb_vorbis include */
+
+#define VVV_MAX_VOLUME 128
+#define VVV_MAX_CHANNELS 16
+
+class SoundTrack;
+class MusicTrack;
+static std::vector<SoundTrack> soundTracks;
+static std::vector<MusicTrack> musicTracks;
+
+static FAudio* faudioctx = NULL;
 
 class SoundTrack
 {
 public:
     SoundTrack(const char* fileName)
     {
-        /* SDL_LoadWAV, convert spec to FAudioBuffer */
         unsigned char *mem;
         size_t length;
+        SDL_AudioSpec spec;
         FILESYSTEM_loadAssetToMemory(fileName, &mem, &length, false);
         if (mem == NULL)
         {
-            m_sound = NULL;
             vlog_error("Unable to load WAV file %s", fileName);
             SDL_assert(0 && "WAV file missing!");
             return;
         }
         SDL_RWops *fileIn = SDL_RWFromConstMem(mem, length);
-        m_sound = Mix_LoadWAV_RW(fileIn, 1);
+        if (SDL_LoadWAV_RW(fileIn, 1, &spec, &wav_buffer, &wav_length) == NULL)
+        {
+            vlog_error("Unable to load WAV file %s", fileName);
+            return;
+        }
         FILESYSTEM_freeMemory(&mem);
-
-        if (m_sound == NULL)
+        format.nChannels = spec.channels;
+        format.nSamplesPerSec = spec.freq;
+        format.wFormatTag = FAUDIO_FORMAT_PCM;
+        format.wBitsPerSample = 16;
+        format.nBlockAlign = format.nChannels * format.wBitsPerSample;
+        format.nAvgBytesPerSec = format.nSamplesPerSec * format.nBlockAlign;
+        format.cbSize = 0;
+
+        if (voices == NULL)
         {
-            vlog_error("Unable to load WAV file: %s", Mix_GetError());
+            voices = (FAudioSourceVoice**)SDL_malloc(sizeof(FAudioSourceVoice*) * VVV_MAX_CHANNELS);
+            for (int i = 0; i < VVV_MAX_CHANNELS; i++)
+            {
+                if (FAudio_CreateSourceVoice(faudioctx, &voices[i], &format, 0, 2.0f, NULL, NULL, NULL))
+                {
+                    vlog_error("Unable to create source voice");
```

Comment:
I don't think this error would be accurate. The voices are only "emulating" the channel retriction. This would be more like "Unable to create source voice no. i")

Reply:
Yeah, you're right.